### PR TITLE
Fix #7088: close AI/GS textfile window when their data are invalid 

### LINKED ISSUE #7088: Crash when viewing AI textfile and switching to Random AI
An example
AIAI version 97 has a changelog.txt

Steps:
- launch openttd
- click AI/GS Settings
- click Select AI
- select AIAI and click Accept
- click Changelog
- without closing Changelog, and without selecting another slot
- click Select AI
- select Random AI and click Accept

Crash!

### PR DESCRIPTION
**Pull request by SamuXarick**
… displaying a textfile

### CODE REVIEW

**Review by LordAro — DISMISSED** [review]
Not a fan of this solution, it's too brittle. The issue boils down to the Window trying to maintain something consistent when the underlying state (the slot) changes.

Instead, the window should "save" the ScriptConfig object (`GetConfig(slot)`) in its constructor, and just refer to that from then on. Should help to fix a couple of the other recent issues (#7090, #7091) as well, I think...

**Comment by SamuXarick** [comment]
I added a solution based on `OnInvalidateData`, to complement the first. The only place I can think of that could alter the content of this window is via `WC_AI_LIST` (the window that lets me chose from a list of available AIs/GSs). So that's where it's called from at the moment.

**Comment by SamuXarick** [comment]
For some reason, this fix stopped working properly. :(

EDIT: fixed it.

File: src/ai/ai_gui.cpp [hunk 1]
```diff
@@ -642,17 +643,32 @@ struct ScriptTextfileWindow : public TextfileWindow {
...
+				const char *script_name = stredup(name);
+				SetDParamStr(1, script_name);
+			}
 		}
 	}
+
+	void OnInvalidateData(int data = 0, bool gui_scope = true) override
+	{
+		const char *textfile = GetConfig(slot)->GetTextfile(file_type, slot);
+		if (textfile == NULL) delete this;
```
Comment (TrueBrain) [inline src/ai/ai_gui.cpp]: No clue what you are smoking here exactly, but this is clearly wrong. Deleting `this` and using it afterwards. That is .. yeah. Clearly wrong.

Comment (SamuXarick) [reply src/ai/ai_gui.cpp]: Thx, that was it.

**Comment by andythenorth** [comment]
Stale.  Closing, thanks.

**Comment by SamuXarick** [comment]
ping for review

File: src/ai/ai_gui.cpp [hunk 2]
```diff
@@ -640,15 +641,33 @@ struct ScriptTextfileWindow : public TextfileWindow {
...
 			SetDParam(0, (slot == OWNER_DEITY) ? STR_CONTENT_TYPE_GAME_SCRIPT : STR_CONTENT_TYPE_AI);
-			SetDParamStr(1, GetConfig(slot)->GetName());
+
+			if (GetConfig(slot)->GetInfo() != nullptr) {
+				SetDParamStr(1, GetConfig(slot)->GetInfo()->GetName());
+			} else {
+				char name[1024];
+				GetString(name, (slot == OWNER_DEITY) ? STR_AI_CONFIG_NONE : STR_AI_CONFIG_RANDOM_AI, lastof(name));
+				const char *script_name = stredup(name);
+				SetDParamStr(1, script_name);
```
Comment (LordAro) [inline src/ai/ai_gui.cpp]: I'm pretty sure there's a memory leak here due to stredup, but is this branch of SetStringParameters even necessary any longer? Given the window will be removed if GetTextfile returns nullptr in OnInvalidateData

**Review by LordAro — DISMISSED** [review]

File: src/ai/ai_gui.cpp (hunk 2 above)
Comment (nielsmh) [reply src/ai/ai_gui.cpp]: I can't tell from just reading this bit of code whether it's possible the window may attempt a repaint with invalid script config/data, but at least it shouldn't be necessary to do the complex string formatting. Just `SetDParamStr(1, "");` and remove the entire GetString and stredup sequence. Nobody's going to see the window with missing text for more that one frame in the worst case.

File: src/ai/ai_gui.cpp [hunk 3]
```diff
@@ -640,15 +641,25 @@ struct ScriptTextfileWindow : public TextfileWindow {
...
+	}
+
+	void OnInvalidateData(int data = 0, bool gui_scope = true) override
+	{
+		const char *textfile = GetConfig(slot)->GetTextfile(file_type, slot);
+		if (textfile == nullptr) {
+			delete this;
+		} else {
+			this->LoadTextfile(textfile, (slot == OWNER_DEITY) ? GAME_DIR : AI_DIR);
+			this->SetStringParameters(WID_TF_CAPTION);
```
Comment (nielsmh) [inline src/ai/ai_gui.cpp]: This line does nothing, it should not be here. `Window::SetStringParameters()` is only intended to be called by the default windowing system drawing code, when it renders a simple control with string data that (might) need string parameters.

**Review by nielsmh — CHANGES_REQUESTED** [review]

**Review by nielsmh — APPROVED** [review]

**Review by LordAro — APPROVED** [review]

**Pull request merged**